okhttp/src/main/java/okhttp3/Challenge.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import java.nio.charset.Charset
+import java.nio.charset.StandardCharsets.ISO_8859_1
+import java.util.Collections.singletonMap
+import java.util.Collections.unmodifiableMap
+import java.util.Locale.US
+
+/** An RFC 7235 challenge. */
+class Challenge(
+  private val scheme: String,
+  authParams: Map<String?, String>
+) {
+  private val authParams: Map<String?, String>
+
+  init {
+    val newAuthParams: MutableMap<String?, String> = LinkedHashMap()
+    for ((key, value) in authParams) {
+      val newKey = key?.toLowerCase(US)
+      newAuthParams[newKey] = value
+    }
+    this.authParams = unmodifiableMap<String?, String>(newAuthParams)
+  }
+
+  constructor(scheme: String, realm: String) : this(scheme, singletonMap("realm", realm))
+
+  /** Returns a copy of this charset that expects a credential encoded with [charset]. */
+  fun withCharset(charset: Charset): Challenge {
+    val authParams = LinkedHashMap(this.authParams)
+    authParams["charset"] = charset.name()
+    return Challenge(scheme, authParams)
+  }
+
+  /** Returns the authentication scheme, like [Basic]. */
+  fun scheme() = scheme
+
+  /**
+   * Returns the auth params, including [realm] and [charset] if present, but as
+   * strings. The map's keys are lowercase and should be treated case-insensitively.
+   */
+  fun authParams() = authParams
+
+  /** Returns the protection space. */
+  fun realm(): String? = authParams["realm"]
```

Comment:
This seems like a breaking change to me and i'm unsure how to proceed. The java code returned a `String` with no `@Nullable` annotation. However Kotlin won't let me do that since it's technically nullable.

Reply:
It's binary compatible but source-incompatible for Kotlin users and NullAway+Java users. I don't think it's a big deal since it'll be attached with a major version. Correctness for nullability is paramount.

Reply:
Should be nullable. This code is right.